sdk/dotnet/Pulumi.Automation/PublicAPI.Unshipped.txt

```diff
@@ -269,6 +269,7 @@ abstract Pulumi.Automation.Workspace.PostCommandCallbackAsync(string stackName,
 abstract Pulumi.Automation.Workspace.Program.get -> Pulumi.Automation.PulumiFn
 abstract Pulumi.Automation.Workspace.Program.set -> void
 abstract Pulumi.Automation.Workspace.PulumiHome.get -> string
+abstract Pulumi.Automation.Workspace.PulumiVersion.get -> Semver.SemVersion
```

Comment:
How important is it to return an instance of `SemVersion` vs. just a string? By doing this, we take the hard dependency on the `Semver` library which is better to avoid if possible. I'd probably go as far as creating our own thin property bag to represent a version rather than exposing the underlying library.

Reply:
Could we use the "Version" type? It's in the System namespace.

Reply:
Every other SDK already has a Semver dependency (as does the `Pulumi` [SDK](https://github.com/pulumi/pulumi/blame/master/sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs#L9) - although I do realize we've split them out for dotnet). So I feel comfortable taking the dependency here. 

I could, as @orionstudt suggested, use `Version` from System instead, but that begs the question why we took the `Semver` dependency in the Pulumi SDK in the first place. And I think you might be the best person to answer that question @mikhailshilkov since you're the one that [added it](https://github.com/pulumi/pulumi/commit/cab39dcfa0f97a35fa526f801b0d0f7d48de40a9#diff-a68acad3699d028399a5705febb7673a0db1f5c664bbdf73e72f5f89629c5187R35)! 🙂 

Additionally, we've exposed them as the SemVer object in every other language as well, so for consistency (and because `SemVersion` is a much more accurate type for this property) I would prefer not to use a string here.

Reply:
I think the built-in `Version` class doesn't support semver v2, so it will fail on our dev/alpha builds.

Taking `SemVer` as an internal dependency is not the same as exposing it in our public API. Right now, we can remove it if we want and nobody notices.

Reply:
> Taking `SemVer` as an internal dependency is not the same as exposing it in our public API.

I'm very sympathetic to this. My general inclination is to avoid exposing another library's types in public APIs, because then our API depends on some external library that we have no control over.

On the other hand, for better or worse, some of our language SDKs have already shipped with such types exposed in their public API, so adding another API that returns the same type in the same SDK makes it no worse than it already is, which was the reasoning behind my comment here: https://github.com/pulumi/pulumi/pull/6577#discussion_r597819438

- Go: https://github.com/pulumi/pulumi/blob/b82b850af0d62461d47ce8948f62ef011c8fdd7b/sdk/go/pulumi/rpc.go#L703-L705
- Python: https://github.com/pulumi/pulumi/blob/768db3e0672225340d2d5e5ea9b9467fa36d6bda/sdk/python/lib/pulumi/runtime/rpc.py#L27 https://github.com/pulumi/pulumi/blob/768db3e0672225340d2d5e5ea9b9467fa36d6bda/sdk/python/lib/pulumi/runtime/rpc.py#L722-L725

What are the scenarios for users calling this API? Having users check versions is generally an anti-pattern, because user's often do the check incorrectly (i.e. they check for specific versions and don't include an "is greater than" clause, causing them to not work correctly on newer versions). The [Windows API for getting the OS's current version](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getversionexw) famously lies in more recent versions of Window (unless the app is manifested) to make existing programs with the problematic version checking work on newer OS versions. Instead, users are encouraged to do feature checking rather than version checking, and if you really have to check the OS version, use [less problematic helpers](https://docs.microsoft.com/en-us/windows/win32/sysinfo/version-helper-apis) instead. Similarly, I don't think we should be encouraging users to check this version. Could we just not expose this API publicly at all (i.e. keep it private and only use it internally)?

Reply:
I think keeping it private sounds like a good way to go, it could just serve internally as a means of verifying that the dependent CLI is an adequate version

Reply:
This was something that was requested by a customer. @justinvp to make sure I'm on the same page, you're now suggesting that I remove the public availability of PulumiVersion from all languages and don't do https://github.com/pulumi/pulumi/issues/6348?

Reply:
Just so everyone's on the same page, in the Go implementation I had initially created "our own thin property bag to represent a version" specifically to avoid taking the dependency. @justinvp asked me to change it to use the SemVer library. So I used that pattern across SDKs.
So I'm a little perplexed at the conversation here, especially since @justinvp has commented on/approved the other PRs.

Reply:
@komal, sorry if my original feedback [here](https://github.com/pulumi/pulumi/pull/6577#discussion_r597819438) led you down a potentially wrong path. I suggested using the existing semver library to do the parsing rather than manual parsing, and for Go, since we already exposed the semver library type elsewhere in public APIs, exposing that rather than creating a custom struct. It wasn't until I saw @mikhailshilkov's questions about returning a string in this PR, that it triggered in my brain that we should be deliberate about whether we want to expose anything (whether it's a custom struct or some external lib's version type) that encourages folks to compare versions, given the problems that it can cause.

> This was something that was requested by a customer.

Do we know what the customer wants to do with the version? If it's just for logging purposes, a string is likely sufficient (and is the same thing the CLI returns), and we wouldn't be encouraging anyone to do comparisons using the API we expose (if a user really wanted to, they could still parse and compare at their own peril).

If the customer wants it to do version checking, are they doing it to see if a specific feature is supported? If so, can we consider alternatives, like exposing feature checking APIs (e.g. `public bool SupportsFoo { get; }`)?

If the custom really is requesting something they can do version checking against, can we consider APIs that encourage more correct behavior, something like (I know it's a mouthful, open to better naming):

```c#
public bool IsPulumiVersionEqualToOrGreaterThan(int major, int? minor = null, int? patch = null, string? prerelease = null);
```

> @justinvp to make sure I'm on the same page, you're now suggesting that I remove the public availability of PulumiVersion from all languages and don't do #6348?

In general, we should be consistent across languages as much as possible (as long as doing so is idiomatic for each language). If we decide not to make it public, it should be private in all languages. If we decide to make it a string, make it a string in all languages.

Reply:
Given we've already shipped this in the other languages via using an external library and we've already exposed the API publicly, it feels like we should be consistent here and just use the library to expose the version.

From a "one-way door" perspective, we've already walked through the door to expose the version. And we've already chosen the way to expose it (via a Semver type in the other languages). Let's not be halfway in the doorway.

Reply:
Sure, but why not use our own Semver type? It will take ~10 lines of code if we are worried about extra work.

Reply:
https://github.com/pulumi/pulumi/pull/6590#discussion_r599726196

Reply:
> since we already exposed the semver library type elsewhere in public APIs, exposing that rather than creating a custom struct.

This does not apply to .NET

Reply:
> It will take ~10 lines of code if we are worried about extra work.

This is a pretty patronizing comment, TBH.

Reply:
Oh, I apologize for inappropriate wording

Reply:
To summarize the feedback, it appears that my options are:
1. Implement a separate `Version` object for C# - because taking the hard dependency on Semver is "better to avoid if possible"
2. Not expose the `PulumiVersion` publicly, and therefore not do #6348 (and change this for the other 3 languages)
3. Expose `PulumiVersion` as a string (and change this for the other 3 languages)

After chatting with the customer in question here and their use-case, it appears 3 is sufficient and what they really want is version validation so we could maybe even get away with doing 2 (assuming that no other customers share the desire to have this accessible). As a happy medium, I will move forward with option 3.

Reply:
Option 3 sounds good to me. Thank you, @komalali!

Reply:
Option 3 is done in this PR for dotnet and in https://github.com/pulumi/pulumi/pull/6604 for the other languages.